Neurodamus building from the new split structure

ferdonline opened the pull request:
Here is a first draft on how to build neurodamus with the new split structure.

I doesn't rely on submodules, but on `dependencies` and `resources`:

- `neurodamus-core` is a dependency, similar to the previous -base. Eventually in the future, neurodamus-core becomes standalone and then it can be renamed to neurodamus.
- `neurodamus-model` has all the logic to build models. Any specific model (e.g. neocortex, hippocampius, thalamus) should inherit from `NeurodamusModel` and basically change the git url. It expects the models to have a `common` folder with common hocs/mods, which shall be a submodule to `sim/models/common`.
In the case where additional hoc/mod files are required (e.g. Thalamus) the module vars `_hoc_srcs` and `_mod_srcs` shall be overriden.

For the moment we are building Nerurodamus "bundles", therefore the naming neurodamus-neocortex, etc... To be better thought the possibility of having standalone models and neurodamus.

ferdonline committed:
Neurodamus building from the new split structure

pramodk commented on var/spack/repos/builtin/packages/neurodamus-base/package.py:
```diff
@@ -30,14 +30,13 @@
 class NeurodamusBase(Package):
```
I saw your comment but I will rename this `sooner`...we will provide these modules for few users to test and ask their opinion. In this case it would be batter to have module `neurodamus-core/0.1` already

ferdonline replied:
Done

pramodk commented on var/spack/repos/builtin/packages/sim-common/package.py:
```diff
@@ -1,39 +1,27 @@
 
-class Neurodamus(NeurodamusBase):
-    """Package used for building special from NeurodamusBase package
+class SimCommon(Package):
```
I am thinking about modules name like:

```
neurodamus-core/0.1
neurodamus-neocortex/0.1
...
```

I am thinking `neurodamus-x` because people are already familiar and there will be less confusion. What do you think?

In that case I would rename `SimCommon` -> `Neurodamus-Common`

ferdonline replied:
my idea was to avoid naming changes again later. Since I expected models to be independent of neurodamus, calling it `neurodamus-cortex` could be misleading. But I also agree the name could have some allusion to neurodamus, I just couldnt come up with a better name. So I thought better to change from the beginning to something meaningful and teach users :) 

Neurodamus-Common since is not really a package, more of a model suplerclass, maybe named to `SimulationModel` ?

ferdonline replied:
I thought a bit and eventually we won't support these models standalone (without neurodamus) - what do you think? Because if that's the case I totally agree with the naming you suggest. I just would like to avoid changing them again some time soon.

ferdonline replied:
Conclusion: For the moment we use the naming neurodamus-xxx. In the future we consider that hard splitting.

pramodk commented on var/spack/repos/builtin/packages/sim-common/package.py:
```diff
@@ -83,19 +67,25 @@ class Neurodamus(NeurodamusBase):
+        copy_tree('hoc', 'hoclib')
+
 
     def build(self, spec, prefix):
```
so if I understood correctly, we will never install `sim-common` explicitly?

ferdonline replied:
exactly... this way we even avoid an installable sim-common, which is somewhat desirable to make clear it's not standalone.

ferdonline replied:
common is now an internal dependency which we handle using submodules.

pramodk commented on var/spack/repos/builtin/packages/sim-common/package.py:
```diff
@@ -83,19 +67,25 @@ class Neurodamus(NeurodamusBase):
+        copy_tree(self.spec['neurodamus-base'].prefix.hoc, 'hoclib')
+        copy_tree('resources/common/hoc', 'hoclib')
+        copy_tree('hoc', 'hoclib')
+
```
why not copy that to hoc and mod instead of hoclib / modlib?

ferdonline replied:
hoc and mod already exist. Maybe poor naming, we can use `merged_hoc`, `merged_mod` instead

ferdonline replied:
Done. Renamed to `_merged_hoc` and `_merged_mod`

pramodk commented on var/spack/repos/builtin/packages/sim-common/package.py:
```diff
@@ -83,19 +67,25 @@ class Neurodamus(NeurodamusBase):
-        os.symlink(self.spec['neurodamus-base'].prefix.lib.modlib, os.path.join(build_dir, 'm'))
+
+    @run_before('build')
+    def merge_hoc_mod(self):
```
I think we should check name conflict...that way it will be safer! you never no when someone will introduce file with same name..!

ferdonline replied:
right. what is you suggestion? by default we have the git clone directly to the temp dir root. Shall we clone to a directory?

ferdonline replied:
For temp libs we use `_merged_xxx` (leading underscore). For the shortname alias, we use `force_symlink`. Hopefully that's enough.

pramodk commented on var/spack/repos/builtin/packages/sim-neocortex/package.py:
```diff
@@ -0,0 +1,14 @@
+from spack.pkg.builtin.sim_common import SimCommon
+
+
+class SimNeocortex(SimCommon):
```
SimNeocortex  -> NeurodamusNeocortex

ferdonline replied:
Done

matz-e commented:
Please test?

matz-e commented:
Please test

Jenkins configuration should hopefully be correct now

ferdonline committed:
Updating neocortex build. Add hippocampus/thalamus

ferdonline committed:
Using submodules instead of resources.

ferdonline committed:
Fixed url -> git

pramodk commented on var/spack/repos/builtin/packages/neurodamus-core/package.py:
```diff
@@ -0,0 +1,21 @@
+    """Library of channels developed by Blue Brain Project, EPFL"""
+
+    homepage = "ssh://bbpcode.epfl.ch/sim/neurodamus-core"
+    git      = "ssh://bbpcode.epfl.ch/sim/neurodamus-core"
```
I would use `url` here

ferdonline replied:
Most packages use git, and that lets spack know we use a git repo by default. Is there any disadvantage?

ferdonline replied:
Since we never distribute tarballs in these projects, I would really prefer `git=` and only that. It makes obvious we only build from the repo, and `version()` will also work even if git=xxx is not provided. As discussed offline, this is now allowed since SPACK PR 3161.

pramodk replied:
fine (minor)

pramodk commented on var/spack/repos/builtin/packages/neurodamus-core/package.py:
```diff
@@ -0,0 +1,21 @@
+    git      = "ssh://bbpcode.epfl.ch/sim/neurodamus-core"
+
+    version('develop', git=git, branch='master')
+    version('2.0.0', tag='2.0.0')
```
any specific reason to start with 2.0.0? I would start with 0.1 :)

ferdonline replied:
Available tags are:
1.3.0
1.3.1
1.3.2
1.6.0
1.8.0
1.9.0

I think it's time to move to V2

ferdonline replied:
Even though we moved (and cleaned) repo, rewriting history, there was no rebase (only condensed merges). So theoretically all tags are 100% the same.
Since last version was 1.9.0, with this important reorganization, I believe we can go 2.0.

pramodk replied:
the question is should keep those tags? with intermediate commits removed/purged, why not start with 0.1? If anyone wants to go back in the history, I would just go to old read-only neurondamus repository. And hence I am thinking to just start from clean : 0.1 :)

ferdonline replied:
I see your point, clean repo, starting from scratch. but if we think about neurodamus as a product, even if we had reimplemented it from scratch it wouldn't go v0.1. Users will think it's a very old and unstable version and avoid to use it.
So my point is, if we keep the version scheme as x.x.x it should be higher than 1.9. Otherwise, with a different scheme (like yyyy.mm) we have full flexibility.

ferdonline replied:
As for the old tags... That's a good question. I brought them, even though it required some more code, so that users could go back to the tag, +/- as if they were in the original repo. 
HOWEVER, even though the branch commits are there, we removed unwanted files from the repo, like tests, and renamed directories. So I totally agree that might confuse the user.
My suggestion is that we give informative names to these tags, but keep them so the developers (us) have still those references (and maybe one day we can rebuild those old versions from the new repo... whatever). They won't hurt

pramodk commented on var/spack/repos/builtin/packages/neurodamus-core/package.py:
```diff
@@ -0,0 +1,21 @@
+        shutil.copytree('hoc', prefix.hoc)
+        shutil.copytree('mod', prefix.mod)
+        if os.path.isdir('python'):
+            shutil.copytree('python', prefix.python)
```
don't we want to set HOC_LIBRARY_PATH?

pramodk replied:
I see now, you combine and set only one path (its fine)

pramodk commented on var/spack/repos/builtin/packages/neurodamus-hippocampus/package.py:
```diff
@@ -0,0 +1,16 @@
+    """
+
+    homepage = "ssh://bbpcode.epfl.ch/sim/models/hippocampus"
+    git      = "ssh://bbpcode.epfl.ch/sim/models/hippocampus"
```
git->url

pramodk commented on var/spack/repos/builtin/packages/neurodamus-hippocampus/package.py:
```diff
@@ -0,0 +1,16 @@
+    git      = "ssh://bbpcode.epfl.ch/sim/models/hippocampus"
+
+    version('develop', git=git, branch='master', submodules=True)
+    version('1.0', tag='1.0', submodules=True)
```
1.0 -> 0.1 ?

ferdonline replied:
After years of devel, these models don't deserve 1.0?

ferdonline replied:
0.1 feels cortex mods are in very early stage. If you are not confident with 1.0 we can go `1.0-alpha`, 2019.01, 0.9.... You name it :)

pramodk replied:
same here...

ferdonline replied:
It's a little detail. I also agree that 1.0 might be too much, sounds very stable. I don't dislike by `yyyy.mm` scheme though. Maybe that's even the best option?

pramodk commented on var/spack/repos/builtin/packages/neurodamus-model/package.py:
```diff
@@ -1,48 +1,31 @@
+    variant('coreneuron', default=False, description="Enable CoreNEURON Support")
+    variant('profile',    default=False, description="Enable profiling using Tau")
+    variant('python',     default=False, description="Enable Python Neurodamus")
+    variant('syntool',    default=True,  description="Enable Synapsetool reader")
```
should we change syntool -> synpasetool for consistency?

ferdonline replied:
Good time to do it now. Thanks!

pramodk commented on var/spack/repos/builtin/packages/neurodamus-model/package.py:
```diff
@@ -1,48 +1,31 @@
+    variant('python',     default=False, description="Enable Python Neurodamus")
+    variant('syntool',    default=True,  description="Enable Synapsetool reader")
+    variant('sonata',     default=False, description="Enable Synapsetool with Sonata")
+    variant('debug',      default=False, description="Compile in debug mode (-O0)")
```
Now I think we should remove debug variant here and just use `spack install neurodamus ^neuron+debug` whenever it requires. What you think?

ferdonline replied:
I introduced it recently because I thought it could make sense to build also the mods as debug. But I'm confused now since I saw mod files being compiled in debug just by using debug Neuron. Whats the magic there? In that case maybe we can drop this yes.

ferdonline replied:
Clarified offline. We can totally drop this option since mod compilation flags are brought from neuron

pramodk commented on var/spack/repos/builtin/packages/neurodamus-model/package.py:
```diff
@@ -1,48 +1,31 @@
+    variant('sonata',     default=False, description="Enable Synapsetool with Sonata")
+    variant('debug',      default=False, description="Compile in debug mode (-O0)")
 
     depends_on("boost", when="+syntool")
```
As we now use chain (and not external synapse tool), I think we can remove boost dependency

pramodk commented on var/spack/repos/builtin/packages/neurodamus-model/package.py:
```diff
@@ -1,48 +1,31 @@
     depends_on("hdf5+mpi")
     depends_on("mpi")
     depends_on("neuron+mpi")
+    depends_on("neuron+mpi+debug", when="+debug")
```
remove this if we decide to remove +debug

pramodk commented on var/spack/repos/builtin/packages/neurodamus-model/package.py:
```diff
@@ -72,30 +51,45 @@ class Neurodamus(NeurodamusBase):
-    conflicts('@hippocampus', when='+coreneuron')
-    conflicts('@master', when='+coreneuron')
+    #conflicts('@hippocampus', when='+coreneuron')
+    #conflicts('@master', when='+coreneuron')
```
you can remove it from here...we will move this in neurodamus-hippocampus

pramodk commented on var/spack/repos/builtin/packages/neurodamus-model/package.py:
```diff
@@ -72,30 +51,45 @@ class Neurodamus(NeurodamusBase):
+    _mod_srcs = ('common/mod', 'mod')
+
+    @staticmethod
+    def copy_all(src, dst, copyfunc=shutil.copy):
```
dst -> dest

matz-e replied:
I like `dst` when one uses `src`…

ferdonline replied:
Same rationale here

pramodk commented on var/spack/repos/builtin/packages/neurodamus-neocortex/package.py:
```diff
@@ -0,0 +1,24 @@
+    """
+
+    homepage = "ssh://bbpcode.epfl.ch/sim/models/neocortex"
+    git      = "ssh://bbpcode.epfl.ch/sim/models/neocortex"
```
same

pramodk commented on var/spack/repos/builtin/packages/neurodamus-neocortex/package.py:
```diff
@@ -0,0 +1,24 @@
+    git      = "ssh://bbpcode.epfl.ch/sim/models/neocortex"
+
+    version('develop', git=git, branch='master', submodules=True)
+    version('1.0', tag='1.0', submodules=True)
```
same

pramodk commented on var/spack/repos/builtin/packages/neurodamus-neocortex/package.py:
```diff
@@ -0,0 +1,24 @@
+    version('develop', git=git, branch='master', submodules=True)
+    version('1.0', tag='1.0', submodules=True)
+
+    variant('v5', default=True, description='Enable support for previous v5 circuits')
```
Do we really need to separate v5 mod files? If so, we might need to make them visible in main mod directory (if Werner and other people ask)

ferdonline replied:
That's why we have a README explaining what we can do to have enerything together (single ln-s )
That's kind of important because (besides cleaning for future, bla, bla, bla) Thalamus requires only V6, and thus this was a simple way to organize the mods.

ferdonline replied:
I hope having the README there is enough, but this separation is valuable I think.

pramodk replied:
Ok, fine for me...lets see what users prefer

pramodk commented on var/spack/repos/builtin/packages/neurodamus-thalamus/package.py:
```diff
@@ -0,0 +1,19 @@
+
+    version('master', git=git, branch='master', submodules=True)
+    version('1.0', tag='1.0', submodules=True)
+
```
same here

ferdonline commented:
Whats the rationale behind renaming git -> url ? Most spack recipes we have use `git` already, and that helps simplifying the `version()` declaration.
According to [packgaing guide](https://spack.readthedocs.io/en/latest/packaging_guide.html#fetching-from-code-repositories) they use `git` for git and `url` for tar.gz installations.

ferdonline committed:
Reviving neurodamus package

Temporarily Support both neurodamus builds in parallel

ferdonline committed:
Reviving old neurodamus p2

ferdonline commented on var/spack/repos/builtin/packages/neurodamus-model/package.py:
```diff
@@ -0,0 +1,164 @@
+
+    def install(self, spec, prefix):
+        """ Move libs to destination.
+            Libs are sym-linked. Compiled libs into libs, special into bin
```
I need also to to through the comments... this is no longer accurate

ferdonline replied:
Done

matz-e commented on var/spack/repos/builtin/packages/neurodamus-base/package.py:
```diff
@@ -0,0 +1,43 @@
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
```
New license is way more compact…

ferdonline replied:
Indeed. I didn't touch those files (`neurodamus-base` and `neurodamus`) though. I had actually deleted them but brought them back since we may have to live with previous neurodamus in parallel for a bit. 
So I would keep them "as is" until we drop them.

ferdonline replied:
@pramodk Or shall we delete right away? and push users to move and we update all tests accordingly?

pramodk replied:
neurodamus-base and neurodamus are dependent...so we have delete them together. But before that we should get green signal from users for new structure...we should send the email to few users tomorrow (to review new structure)

ferdonline committed:
Addressed comments. Droppped Python support

Addressed a lot of dependency simplifications, namely due to the fact
that we dont require to suport external static synapsetool and python is
no longer part of core
Some naming changed.

ferdonline committed:
Fix typo

pramodk commented on var/spack/repos/builtin/packages/neurodamus-model/package.py:
```diff
@@ -0,0 +1,138 @@
+    # Note: With Spack chain we no longer require support for external libs
+    # which could bring dependencoies. However, in some setups (notably tests)
+    # Some libraries are still external so we bring these dependencies
+    depends_on('zlib')  # for hdf5
```
I assume this will go away once we switch chain as well to chain

pramodk commented on var/spack/repos/builtin/packages/neurodamus-model/package.py:
```diff
@@ -0,0 +1,138 @@
+    depends_on('reportinglib+profile', when='+profile')
+    depends_on('tau', when='+profile')
+
+    # ---
```
`# ---` can be removed?

pramodk reviewed (commented):
@ferdonline  : overall looks fine to me...lets quickly sync versioning issue tomorrow morning.

ferdonline committed:
Adding neurodamus-neocortex to module whitelist

ferdonline committed:
Add Python and Plasticity support

ferdonline committed:
Use models/neocortex v 1.1

matz-e commented on var/spack/repos/builtin/packages/neurodamus-model/package.py:
```diff
@@ -0,0 +1,158 @@
+    # Some libraries are still external so we bring these dependencies
+    depends_on('zlib')  # for hdf5
+
+    # Profiling
```
Isn't this obvious/redundant from the variant name?

matz-e reviewed (approved):
Looks good to me.

ferdonline committed:
Addressed review. Better comments

pramodk commented:
@ferdonline  : I build this on my mac and able to use modules / run neocortex module. So this is fine to merge. (note that while merging we should do "Squash & Merge" and clean-up commit message if necessary)

ferdonline commented:
To get things moving, merging now. We can deal with hippocampus and thalamus issues subsequently. 
As for the versions let's keep as is. In the end neurodamus-core version is not publicly visible, only the neurodamus-xxx have versions which are starting at 1.0
